// keyboard-simulator/dist/esm/keyboard-simulator.esm.js
/*! keyboard-simulator v0.0.10 | MIT License | © Itai Tenenbaum 2026 | https://github.com/taitulism/keyboard-simulator */
const TogglerButtons = ['NumLock', 'CapsLock', 'ScrollLock'];
const isAlias = (key) => key in KeyAliases;
const isKeyId = (key) => key in KeyMap;
const isModifier = (str) => str in ModifierNumbers;
const isTogglerBtn = (str) => TogglerButtons.includes(str);
const isAffectedByNumLock = (keyId) => /Numpad\d/.test(keyId) || keyId.endsWith('Decimal');
const isAffectedByCapsLock = (keyId) => keyId in LettersKeyMap;
const isAffectedByShift = (keyId) => keyId in LettersKeyMap || keyId in NumbersKeyMap || keyId in SymbolsKeyMap;
const getKeyId = (keyName) => {
    if (isKeyId(keyName))
        return keyName;
    const lower = keyName.toLowerCase();
    if (isAlias(lower))
        return KeyAliases[lower];
    throw new Error(`Unknown key name: ${keyName}`);
};
const getKeyValue = (keyId, isAlterValue) => {
    // value = single or array
    const value = KeyMap[keyId];
    if (Array.isArray(value)) {
        return isAlterValue
            ? value[1]
            : value[0];
    }
    return value;
};
const ModifierNumbers = {
    ControlLeft: 1,
    ControlRight: 1,
    AltLeft: 2,
    AltRight: 2,
    ShiftLeft: 3,
    ShiftRight: 3,
    MetaLeft: 4,
    MetaRight: 4,
};
const SymbolsKeyMap = {
    Slash: ['/', '?'],
    Backslash: ['\\', '|'],
    IntlBackslash: ['\\', '|'],
    Period: ['.', '>'],
    Comma: [',', '<'],
    Quote: ['\'', '"'],
    Backquote: ['`', '~'],
    Semicolon: [';', ':'],
    BracketLeft: ['[', '{'],
    BracketRight: [']', '}'],
    Minus: ['-', '_'],
    Equal: ['=', '+'],
};
const LettersKeyMap = {
    KeyA: ['a', 'A'],
    KeyB: ['b', 'B'],
    KeyC: ['c', 'C'],
    KeyD: ['d', 'D'],
    KeyE: ['e', 'E'],
    KeyF: ['f', 'F'],
    KeyG: ['g', 'G'],
    KeyH: ['h', 'H'],
    KeyI: ['i', 'I'],
    KeyJ: ['j', 'J'],
    KeyK: ['k', 'K'],
    KeyL: ['l', 'L'],
    KeyM: ['m', 'M'],
    KeyN: ['n', 'N'],
    KeyO: ['o', 'O'],
    KeyP: ['p', 'P'],
    KeyQ: ['q', 'Q'],
    KeyR: ['r', 'R'],
    KeyS: ['s', 'S'],
    KeyT: ['t', 'T'],
    KeyU: ['u', 'U'],
    KeyV: ['v', 'V'],
    KeyW: ['w', 'W'],
    KeyX: ['x', 'X'],
    KeyY: ['y', 'Y'],
    KeyZ: ['z', 'Z'],
};
const NumbersKeyMap = {
    Digit1: ['1', '!'],
    Digit2: ['2', '@'],
    Digit3: ['3', '#'],
    Digit4: ['4', '$'],
    Digit5: ['5', '%'],
    Digit6: ['6', '^'],
    Digit7: ['7', '&'],
    Digit8: ['8', '*'],
    Digit9: ['9', '('],
    Digit0: ['0', ')'],
};
const KeyMap = {
    ...LettersKeyMap,
    ...NumbersKeyMap,
    ...SymbolsKeyMap,
    // NumPad (Numlock)
    Numpad0: ['Insert', '0'],
    Numpad1: ['End', '1'],
    Numpad2: ['ArrowDown', '2'],
    Numpad3: ['PageDown', '3'],
    Numpad4: ['ArrowLeft', '4'],
    Numpad5: ['Clear', '5'],
    Numpad6: ['ArrowRight', '6'],
    Numpad7: ['Home', '7'],
    Numpad8: ['ArrowUp', '8'],
    Numpad9: ['PageUp', '9'],
    NumpadDecimal: ['Delete', '.'],
    NumpadDivide: '/',
    NumpadSubtract: '-',
    NumpadMultiply: '*',
    NumpadAdd: '+',
    // Arrows / Navigation
    ArrowUp: 'ArrowUp',
    ArrowRight: 'ArrowRight',
    ArrowDown: 'ArrowDown',
    ArrowLeft: 'ArrowLeft',
    PageUp: 'PageUp',
    PageDown: 'PageDown',
    Home: 'Home',
    End: 'End',
    // Modifiers
    ControlLeft: 'Control',
    ControlRight: 'Control',
    AltLeft: 'Alt',
    AltRight: 'Alt',
    ShiftLeft: 'Shift',
    ShiftRight: 'Shift',
    MetaLeft: 'Meta',
    MetaRight: 'Meta',
    // Text Spaces
    Insert: 'Insert',
    Delete: 'Delete',
    Enter: 'Enter',
    NumpadEnter: 'Enter',
    Space: ' ',
    Backspace: 'Backspace',
    Tab: 'Tab',
    // Fn
    F1: 'F1',
    F2: 'F2',
    F3: 'F3',
    F4: 'F4',
    F5: 'F5',
    F6: 'F6',
    F7: 'F7',
    F8: 'F8',
    F9: 'F9',
    F10: 'F10',
    F11: 'F11',
    F12: 'F12',
    F13: 'F13',
    F14: 'F14',
    F15: 'F15',
    F16: 'F16',
    F17: 'F17',
    F18: 'F18',
    F19: 'F19',
    F20: 'F20',
    F21: 'F21',
    F22: 'F22',
    F23: 'F23',
    F24: 'F24',
    // Others
    Pause: 'Pause',
    PrintScreen: 'PrintScreen',
    ScrollLock: 'ScrollLock',
    NumLock: 'NumLock',
    CapsLock: 'CapsLock',
    ContextMenu: 'ContextMenu',
    Escape: 'Escape',
};
/* ======================================================= */
// All keys are lowercased
const KeyAliases = {
    // Letters
    a: 'KeyA',
    b: 'KeyB',
    c: 'KeyC',
    d: 'KeyD',
    e: 'KeyE',
    f: 'KeyF',
    g: 'KeyG',
    h: 'KeyH',
    i: 'KeyI',
    j: 'KeyJ',
    k: 'KeyK',
    l: 'KeyL',
    m: 'KeyM',
    n: 'KeyN',
    o: 'KeyO',
    p: 'KeyP',
    q: 'KeyQ',
    r: 'KeyR',
    s: 'KeyS',
    t: 'KeyT',
    u: 'KeyU',
    v: 'KeyV',
    w: 'KeyW',
    x: 'KeyX',
    y: 'KeyY',
    z: 'KeyZ',
    // Numbers
    '0': 'Digit0',
    '1': 'Digit1',
    '2': 'Digit2',
    '3': 'Digit3',
    '4': 'Digit4',
    '5': 'Digit5',
    '6': 'Digit6',
    '7': 'Digit7',
    '8': 'Digit8',
    '9': 'Digit9',
    np0: 'Numpad0',
    np1: 'Numpad1',
    np2: 'Numpad2',
    np3: 'Numpad3',
    np4: 'Numpad4',
    np5: 'Numpad5',
    np6: 'Numpad6',
    np7: 'Numpad7',
    np8: 'Numpad8',
    np9: 'Numpad9',
    decimal: 'NumpadDecimal',
    divide: 'NumpadDivide',
    subtract: 'NumpadSubtract',
    multiply: 'NumpadMultiply',
    add: 'NumpadAdd',
    // Arrows / Navigation
    up: 'ArrowUp',
    right: 'ArrowRight',
    down: 'ArrowDown',
    left: 'ArrowLeft',
    pgup: 'PageUp',
    pgdn: 'PageDown',
    // Modifiers
    control: 'ControlLeft',
    ctrl: 'ControlLeft',
    lctrl: 'ControlLeft',
    rctrl: 'ControlRight',
    alt: 'AltLeft',
    lalt: 'AltLeft',
    ralt: 'AltRight',
    shift: 'ShiftLeft',
    lshift: 'ShiftLeft',
    rshift: 'ShiftRight',
    meta: 'MetaLeft',
    lmeta: 'MetaLeft',
    rmeta: 'MetaRight',
    // Symbols
    singlequote: 'Quote',
    backtick: 'Backquote',
    // Enter / Spaces
    npenter: 'NumpadEnter',
    renter: 'NumpadEnter',
    del: 'Delete',
    ins: 'Insert',
    // Others
    esc: 'Escape',
};

const isObject = (thing) => typeof thing === 'object' && thing !== null;
const isDocument = (doc) => isObject(doc) && 'nodeType' in doc && doc.nodeType === 9; // Node.DOCUMENT_NODE = 9
// doc.constructor.name === 'HTMLDocument'; // fails in JSDOM ('Document')
class KeyboardSimulator {
    constructor(ctxElm) {
        this.isCtrlDown = false;
        this.isAltDown = false;
        this.isShiftDown = false;
        this.isMetaDown = false;
        this.isCapsLockOn = false;
        this.isScrollLockOn = false;
        this.isNumLockOn = true;
        this.heldKeys = new Set();
        this._ctxElm = null;
        this.setContextElm(ctxElm);
    }
    set doc(doc) {
        this._doc = doc;
    }
    get doc() {
        if (!this._doc)
            this._doc = document; // Set once (global)
        return this._doc;
    }
    get ctxElm() {
        if (this._ctxElm)
            return this._ctxElm;
        const { doc } = this; // Getter activated
        // TODO:ts "as HTMLElement" is wrong. Should be "Element" (svg & iframe)
        if (doc.activeElement)
            return doc.activeElement;
        return doc;
    }
    setContextElm(ctxElm) {
        if (!ctxElm) {
            this._ctxElm = null;
        }
        else if (isDocument(ctxElm)) {
            this.doc = ctxElm;
        }
        else { // is HTMLElement
            if (!this._doc) {
                this.doc = ctxElm.ownerDocument;
            }
            this._ctxElm = ctxElm;
        }
        return this;
    }
    followKey(key) {
        if (this.heldKeys.has(key))
            throw new Error(`The key "${key}" is already pressed down.`);
        this.heldKeys.add(key);
    }
    unfollowKey(key) {
        if (!this.heldKeys.has(key))
            throw new Error(`The key "${key}" is not pressed down.`);
        this.heldKeys.delete(key);
    }
    reset() {
        this.isCtrlDown = false;
        this.isAltDown = false;
        this.isShiftDown = false;
        this.isMetaDown = false;
        this.isCapsLockOn = false;
        this.isNumLockOn = true;
        this.isScrollLockOn = false;
        this.heldKeys.clear();
        this._ctxElm = null;
    }
    toggleModifier(keyId, isPressed = false) {
        const modifier = ModifierNumbers[keyId];
        if (modifier === 1)
            this.isCtrlDown = isPressed;
        else if (modifier === 2)
            this.isAltDown = isPressed;
        else if (modifier === 3)
            this.isShiftDown = isPressed;
        else if (modifier === 4)
            this.isMetaDown = isPressed;
    }
    toggleToggler(togglerBtn) {
        if (togglerBtn === 'NumLock')
            this.isNumLockOn = !this.isNumLockOn;
        else if (togglerBtn === 'CapsLock')
            this.isCapsLockOn = !this.isCapsLockOn;
        else if (togglerBtn === 'ScrollLock')
            this.isScrollLockOn = !this.isScrollLockOn;
    }
    keyDown(...keys) {
        const dispatchResults = keys.map((keyName) => {
            const keyId = getKeyId(keyName);
            this.followKey(keyId);
            if (isModifier(keyId))
                this.toggleModifier(keyId, true);
            else if (isTogglerBtn(keyId))
                this.toggleToggler(keyId);
            const keyDownEvent = this.createKeyboardEvent('keydown', keyId);
            return this.ctxElm.dispatchEvent(keyDownEvent);
        });
        return (dispatchResults.length === 1) ? dispatchResults[0] : dispatchResults;
    }
    keyUp(...keys) {
        const dispatchResults = keys.map((keyName) => {
            const keyId = getKeyId(keyName);
            this.unfollowKey(keyId);
            if (isModifier(keyId))
                this.toggleModifier(keyId, false);
            const keyUpEvent = this.createKeyboardEvent('keyup', keyId);
            return this.ctxElm.dispatchEvent(keyUpEvent);
        });
        return (dispatchResults.length === 1) ? dispatchResults[0] : dispatchResults;
    }
    keyPress(...keys) {
        const dispatchResults = keys.map((key) => [
            this.keyDown(key),
            this.keyUp(key),
        ]);
        return (dispatchResults.length === 1) ? dispatchResults[0] : dispatchResults;
    }
    combine(...keys) {
        const dispatchDownResults = keys.map((key) => this.keyDown(key));
        const dispatchUpResults = keys.reverse().map((key) => this.keyUp(key));
        return [dispatchDownResults, dispatchUpResults];
    }
    createKeyboardEvent(eventType, keyName, eventInit = {}) {
        const keyId = getKeyId(keyName);
        const isAlternativeValue = this.isNumLockOn && isAffectedByNumLock(keyId)
            || this.isShiftDown && isAffectedByShift(keyId)
            || this.isCapsLockOn && isAffectedByCapsLock(keyId);
        const keyValue = getKeyValue(keyId, isAlternativeValue);
        const initialObj = {
            code: keyId,
            key: keyValue,
            ctrlKey: this.isCtrlDown,
            altKey: this.isAltDown,
            shiftKey: this.isShiftDown,
            metaKey: this.isMetaDown,
            repeat: false,
            location: 1,
            bubbles: true,
            cancelable: true,
            composed: true, // event will propagate across the shadow DOM boundary into the standard DOM
            isComposing: false, // For Emojis and other special characters
            view: this.doc.defaultView,
        };
        const eventObj = Object.assign(initialObj, eventInit);
        return new KeyboardEvent(eventType, eventObj);
    }
    repeat(count) {
        if (!this.heldKeys.size)
            throw new Error('Cannot repeat. No keys are pressed down.');
        const lastKey = Array.from(this.heldKeys)[this.heldKeys.size - 1];
        const dispatches = [];
        const opts = { repeat: true };
        for (let i = 0; i < count; i++) {
            const keyDownEvent = this.createKeyboardEvent('keydown', lastKey, opts);
            dispatches.push(this.ctxElm.dispatchEvent(keyDownEvent));
        }
        return dispatches;
    }
    release(...keys) {
        if (keys.length) {
            return this.keyUp(...keys);
        }
        const dispatches = Array.from(this.heldKeys)
            .reverse()
            .map((key) => this.keyUp(key));
        this.heldKeys.clear();
        return (dispatches.length === 1) ? dispatches[0] : dispatches;
    }
}

export { KeyAliases, KeyMap, KeyboardSimulator, ModifierNumbers, getKeyId, getKeyValue, isAffectedByCapsLock, isAffectedByNumLock, isAffectedByShift, isModifier, isTogglerBtn };
